### DISCUSSION TITLE
[Python] Add dynamic attributes to DuckDBPyRelation instances

### DISCUSSION THREAD

**Discussion by Tishj** [duckdb/duckdb#6468] [Ideas]
Pandas has dynamic properties on a DataFrame instance that expose an internal Series object.

Through the `__dir__` method, this allows IDE's to support auto-completion/suggestion.
Pandas properties have getters and setters, so this allows you to alter the DataFrame by setting.
We're only interested in the getter, creating a projection relation on the existing dataframe of the column that was asked for.

**Comment by Tishj** [comment]
I'm looking into this, and from what I read it's done through the `__dir__` method, to check which attributes an instance has.

I'm not having the best success verifying that that's how this is done, as
```py
df = pd.DataFrame({'a': [5,4,3]})
df.__dir__()
```
does not contain a key named `a`

But regardless, the main issue with implementing this is that we need more granular control over the python object instance, which is now managed purely by pybind11.
We'd need some sort of middle-ware method that receives the created python object from the `shared_ptr<DuckDBPyRelation>` and then lets us define properties on it based on the columns it contains

**Reply by Mause** [reply]
Sounds like we need to add a `__getattr__` method to the relation that looks up the appropriate name on the DataFrame at call time? That way we don't need to know ahead of time

**Reply by Tishj** [reply]
Do you think that will also be enough to get auto-complete/suggestion in IDEs?

**Reply by Tishj** [reply]
Added `__getattr__` here:
https://github.com/tishj/duckdb/tree/pyrelation_getattr
